### PR TITLE
Improve Alliance and Peace Logic

### PR DESCRIPTION
**Pull request by RunzheYang**
- include new stance modules (default betrayal flip/ cleaned stance log / force update stance)
 - support PCE parsing and response
 - REJ ALY/PCE from powers with stance scores below the threshold
-  update stance and allies after replying to ALY/PCE

### CODE REVIEW

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (aphedges) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:864]: Do you mean display the updated stance explanation here?

Comment (RunzheYang) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:864]: do you mean to report stance scores in GLOBAL messages? 
stance scores are currently in intent logging.

Comment (aphedges) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:864]: Also, please replace this GLOBAL message with intent logging like was done earlier in the function. This should all go into intent logging that that it's an option.

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (ezubaric) [inline src/baseline_bots/bots/smart_order_accepter_bot.py:491]: I'd feel better about this if there were an assert here to check that there was no overlap.

File: src/baseline_bots/utils.py
Comment (ezubaric) [inline src/baseline_bots/utils.py:256]: In a perfect world, this code could be shared with alliance for less duplication, but probably don't have time for that

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (ezubaric) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:373]: So this is saying that they haven’t attacked you more than N times?

Comment (RunzheYang) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:373]: this filters out the case where the opponent is extremely hostile. 
it's multiplicative because the discount factor is multiplicative

Comment (ezubaric) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:373]: But regardless, should be a parameter and distinct for peace and alliance

**Review by ezubaric — APPROVED** [review]
Some comments, not requiring changes (we need to test), but would be good to do

File: src/baseline_bots/utils.py
Comment (aphedges) [reply src/baseline_bots/utils.py:256]: Later this week, when the deadline is met, I'm going to spend a lot of time doing code cleanup and refactoring. The parsing code will be one of my focuses.

**Review by aphedges — CHANGES_REQUESTED** [review]
The GLOBAL code needs to be changed to intent logging, so this shouldn't be merged yet.

I'll approve after @RunzheYang makes his changes and I do cleanup following that.

File: src/baseline_bots/utils.py
Comment (aphedges) [reply src/baseline_bots/utils.py:256]: I have been looking into an existing DAIDE-focused parsing library and plan to bring it up to the team as a whole soon.

**Comment by aphedges** [comment]
> > The GLOBAL code needs to be changed to intent logging, so this shouldn't be merged yet.
> > I'll approve after @RunzheYang makes his changes and I do cleanup following that.
> 
> @aphedges what changes need to be made here? I kept your change to intent logging and also threw the final stance massages to GLOBAL. I think we need to post the stance in GLOBAL?

There was a discussion in Slack yesterday that it should be changed to intent logging. @ezubaric should correct me if I misunderstood.

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (RunzheYang) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:373]: ok will add them.

**Comment by RunzheYang** [comment]
> > > The GLOBAL code needs to be changed to intent logging, so this shouldn't be merged yet.
> > > I'll approve after @RunzheYang makes his changes and I do cleanup following that.
> > 
> > 
> > @aphedges what changes need to be made here? I kept your change to intent logging and also threw the final stance massages to GLOBAL. I think we need to post the stance in GLOBAL?
> 
> There was a discussion in Slack yesterday that it should be changed to intent logging. @ezubaric should correct me if I misunderstood.

OK I'll remove that then.

**Comment by aphedges** [comment]
@RunzheYang, when you've made all the changes discussed, please re-request review from me so I can clean this up and merge.

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (ezubaric) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:864]: But scores here with the sets (if that's easy): "%s %0.2f" % (power, score) or something like that.

Comment (aphedges) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:864]: That makes sense! Given that this block will be deleted, I'll modify the one we are keeping.

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (aphedges) [inline src/baseline_bots/bots/smart_order_accepter_bot.py:342]: Make this print statement an intent log as well.

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (aphedges) [inline src/baseline_bots/bots/smart_order_accepter_bot.py:354]: Instead of printing this, add this to the intent logger. I'm making a change to send all intent logging to console for easier debugging.

File: tests/bot_tests/SOA_test.py
Comment (aphedges) [inline tests/bot_tests/SOA_test.py:186]: Can this assert be fixed? If not, at least remove the print statement, as it's not needed for the test.

**Review by aphedges — CHANGES_REQUESTED** [review]

File: src/baseline_bots/bots/smart_order_accepter_bot.py
Comment (RunzheYang) [reply src/baseline_bots/bots/smart_order_accepter_bot.py:491]: ok added.

File: tests/bot_tests/SOA_test.py
Comment (RunzheYang) [reply tests/bot_tests/SOA_test.py:186]: output can be REJ now so not sure how to fix this assert... this test at least makes sure no running time issue for getting responses.

Comment (aphedges) [reply tests/bot_tests/SOA_test.py:186]: Then can you please remove the `print` statement? That output will not help the test in any way.

**Review by aphedges — CHANGES_REQUESTED** [review]

**Review by aphedges — APPROVED** [review]
This all looks good to me! I've squashed and rebased the branch and wrote a cohesive commit message. Once CI passes, I will merge this.

**Pull request merged**
